python/ray/dag/tests/experimental/test_cpu_communicator_dag.py

```diff
@@ -0,0 +1,375 @@
+import os
+import sys
+import torch
+
+import pytest
+
+import ray
+import ray.cluster_utils
+from ray.exceptions import RayChannelError
+from ray.experimental.channel.torch_tensor_type import TorchTensorType
+from ray.experimental.channel.cpu_nccl_group import CPUNcclGroup, start_nccl_mock
+from ray.dag import InputNode
+import ray.experimental.collective as collective
+from ray.dag.output_node import MultiOutputNode
+from ray.tests.conftest import *  # noqa
+
+
+@ray.remote
+class CPUTorchTensorWorker:
+    def start_cuda_and_torch_mock(self):
+        start_nccl_mock()
+
+    def send(self, shape, dtype, value: int, send_tensor=True):
+        if not send_tensor:
+            return 1
+        return torch.ones(shape, dtype=dtype) * value
+
+    def send_dict(self, entries):
+        results = {}
+        for key, entry in entries.items():
+            value, shape, dtype = entry
+            results[key] = torch.ones(shape, dtype=dtype) * value
+        return results
+
+    def send_or_raise(self, shape, dtype, value: int, raise_exception=False):
+        if raise_exception:
+            raise RuntimeError()
+        return torch.ones(shape, dtype=dtype) * value
+
+    def recv(self, tensor):
```

Comment:
When the tensor is received, its device would appear to be ‘cuda’ even though the test environment does not have a GPU. This is because of `ctx.set_torch_device(torch.device(“cuda”))` in start_nccl_mock — however, if we remove this, there are upstream assertions in torch_tensor_nccl_channel.py that fail (for example, assert ctx.torch_device.type == “cuda” in ensure_registered_as_writer). We wanted to ask what the best way to fix this issue was — maybe some custom logic in ensure_registered_as_writer to check if a CPUNcclGroup is being used and not fail in that case?

Reply:
Hmm I don't think we should be using the mock in this test. The mock should only be needed for cases where a nccl group is used; the CPU group should be usable without the mock, right?

Possibly you just need to update that assertion check / move it to the NCCL group code instead.

Reply:
fixed